test/test_invalid_headers.py

```diff
@@ -95,30 +97,49 @@ class TestFilter(object):
     HTTP/2 and so may never hit the function, but it's worth validating that it
     behaves as expected anyway.
     """
-    true_false_combinations = [
-        (True, True),
-        (True, False),
-        (False, True),
-        (False, False),
+    validation_functions = [
+        h2.utilities.validate_headers,
+        h2.utilities.validate_sent_headers
     ]
+    validator_combinations = list(itertools.product(
+        validation_functions,
+        (True, False),
+        (True, False),
+    ))
 
-    @pytest.mark.parametrize('is_client,is_trailer', true_false_combinations)
+    @pytest.mark.parametrize('validation_function,is_client,is_trailer',
+                             validator_combinations)
     @given(headers=HEADERS_STRATEGY)
-    def test_range_of_acceptable_outputs(self, headers, is_client, is_trailer):
+    def test_range_of_acceptable_outputs(self,
+                                         headers,
+                                         validation_function,
+                                         is_client,
+                                         is_trailer):
         """
-        validate_headers either returns the data unchanged or throws a
-        ProtocolError.
+        The header validation functions either return the data unchanged
+        or throw a ProtocolError.
         """
         hdr_validation_flags = h2.utilities.HeaderValidationFlags(
             is_client=is_client,
             is_trailer=is_trailer,
         )
         try:
-            assert headers == h2.utilities.validate_headers(
-                headers, hdr_validation_flags)
+            assert headers == list(validation_function(
+                headers, hdr_validation_flags))
```

Comment:
Possibly this is the wrong fix; this was an attempt to fix a failure when a client tries to send an empty list of headers which aren't trailers:

```
        try:
>           assert headers == validation_function(
                headers, hdr_validation_flags)
E               assert [] == <generator object _validate_...thority_header at 0x279b910>
E                 Full diff:
E                 - []
E                 + <generator object _validate_host_authority_header at 0x279b910>

test/test_invalid_headers.py:127: AssertionError
------------------------------------------------------------------------- Hypothesis -------------------------------------------------------------------------
Falsifying example: test_range_of_acceptable_outputs(self=<test_invalid_headers.TestFilter at 0x28f0510>, headers=[], validation_function=validate_sent_headers, is_client=True, is_trailer=False)
============================================================ 1 failed, 31 passed in 3.04 seconds =============================================================
```